File: src/simulate.jl
function step(model, x, u; h::Float64=0.01, grouped_aero_params=nothing, printing=false)
    ẋ = dynamics(model, x, u; printing=printing, grouped_aero_params=grouped_aero_params)
    xn = rk4(x, x_ -> dynamics(model, x_, u; grouped_aero_params=grouped_aero_params), h)
    if model isa QuadRotor && model.ground_check && xn[3] > 0
        xn[3] = 0
    end
    return xn, ẋ
end


function simulate(model, x0, us, time::Number; h::Float64=0.01, controller=nothing, grouped_aero_params=nothing, printing=false)
    t = 0
    i = 1
    xs = [x0]
    ẋs = []
    us_ = []
    while t < time
        if isnothing(controller)
            u = us[i]
        else
            u = QuadrotorDynamics.get_control(controller, xs[end], t)
        end
        x, ẋ = step(model, xs[end], u; h=h, printing=printing, grouped_aero_params=grouped_aero_params)
        push!(xs, x)
        push!(ẋs, ẋ)
        push!(us_, u)
        t += h
        i += 1
    end
    return xs, ẋs, us_
end

File: python/simulate.py
import numpy as np
from utils import rk4


def step(quad, x, u, h, z=None):
    if z is None:
        xn = rk4(x, lambda _x: quad.dynamics(_x, u)[0], h)
        zn = None
    else:
        xzn = rk4(hstack((x, z)), lambda xz: hstack(quad.dynamics(xz[:quad.xdim], u, xz[quad.xdim:])), h)
        xn, zn = torch.split(xzn, [quad.xdim, quad.zdim])
    return xn, zn


def simulate(quad, x0, us, time, h, z0=None):
    t = i = 0
    xs = [x0]
    zs = [z0]
    for i, t in enumerate(np.arange(0, time + 1e-7, h)):
        xn, zn = step(quad, xs[-1], us[i], h, zs[-1])
        xs.append(xn)
        zs.append(zn)
    return xs, zs
